dataflow/src/main/java/org/checkerframework/dataflow/cfg/builder/CFGTranslationPhaseOne.java

```diff
@@ -3602,172 +3607,344 @@ public Node visitTry(TryTree tree, Void p) {
     }
 
     if (finallyLabel != null) {
-      // Reset values before analyzing the finally block!
+      handleFinally(
+          tree,
+          doneLabel,
+          finallyLabel,
+          exceptionalFinallyLabel,
+          () -> scan(finallyBlock, p),
+          oldReturnTargetLC,
+          oldBreakTargetLC,
+          oldBreakLabels,
+          oldContinueTargetLC,
+          oldContinueLabels);
+    }
 
-      tryStack.popFrame();
+    addLabelForNextNode(doneLabel);
 
-      { // Scan 'finallyBlock' for only 'finallyLabel' (a successful path)
-        addLabelForNextNode(finallyLabel);
-        extendWithNode(
-            new MarkerNode(
-                tree,
-                "start of finally block #" + TreeUtils.treeUids.get(tree),
-                env.getTypeUtils()));
-        scan(finallyBlock, p);
-        extendWithNode(
-            new MarkerNode(
-                tree, "end of finally block #" + TreeUtils.treeUids.get(tree), env.getTypeUtils()));
-        extendWithExtendedNode(new UnconditionalJump(doneLabel));
-      }
+    return null;
+  }
 
-      if (hasExceptionalPath(exceptionalFinallyLabel)) {
-        // If an exceptional path exists, scan 'finallyBlock' for 'exceptionalFinallyLabel',
-        // and scan copied 'finallyBlock' for 'finallyLabel' (a successful path). If there
-        // is no successful path, it will be removed in later phase.
-        // TODO: Don't we need a separate finally block for each kind of exception?
-        addLabelForNextNode(exceptionalFinallyLabel);
-        extendWithNode(
-            new MarkerNode(
-                tree,
-                "start of finally block for Throwable #" + TreeUtils.treeUids.get(tree),
-                env.getTypeUtils()));
+  /**
+   * Helper function to handle the resource declarations in a {@link TryTree} and its main block.
+   * The logic here follows JLS 14.20.3.1. A resource declaration <i>r</i> is handled by adding the
+   * nodes for <i>r</i> itself to the CFG, followed by a simulated nested try block <i>t1</i> for
+   * any remaining resource declarations and the original try block (handled with a recursive call).
+   * A call to {@code close} for <i>r</i> is added in a simulated "finally" block for <i>t1</i>,
+   * guaranteeing that on every path through the CFG, <i>r</i> is closed.
+   *
+   * @param tryTree the original try tree from the AST
+   * @param p value to pass to calls to {@code scan}
+   * @param resources remaining resource declarations to handle
+   */
+  private void visitTryResourcesAndBodyHelper(
```

Comment:
I agree the naming here is not ideal.  I tried to clarify the Javadoc to indicate that this method is used for any `TryTree`, whether it has resource declarations or not.  For a case with no resource declarations, the `resources` parameter will be empty, and the base case logic will run, scanning the block of the try.

I named the method `visitTryResourcesAndBodyHelper` to try to capture how it first handles the resource declarations (if any) and then the body.  Would `visitTryResourcesAndBlockHelper` be better?

Reply:
Maybe it would help to move "helper" to the beginning of the method name.  Starting with "visit" makes this seem like a visitor, and then I tried to interpret the rest of the method name as what was being visited.

I'm OK with the current name, too, though, if you think all alternatives would be worse.

Reply:
I went with `handleTryResourcesAndBlock` to avoid confusion regarding "visit" and to refer to the block rather than the body.